File: android/app/src/main/java/com/guardforce/guard/data/sync/Payloads.kt
package com.guardforce.guard.data.sync

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

object Kinds {
    const val CHECK_IN = "CHECK_IN"
    const val CHECK_OUT = "CHECK_OUT"
    const val LOCATION_STATE = "LOCATION_STATE"
    const val PATROL_START = "PATROL_START"
    const val PATROL_COMPLETE = "PATROL_COMPLETE"
    const val TASK_START = "TASK_START"
    const val TASK_COMPLETE = "TASK_COMPLETE"
    const val LEAVE_APPLY = "LEAVE_APPLY"
    const val LEAVE_CANCEL = "LEAVE_CANCEL"
    const val REG_SELFIE = "REG_SELFIE"
}

@Serializable data class LocalPhoto(val file: String, val remotePath: String, val lat: Double? = null, val lng: Double? = null, val takenAt: String)

@Serializable
data class CheckInPayload(
    val shiftKey: String, val guardId: String, val siteId: String,
    val lat: Double, val lng: Double, val accuracyM: Float?,
    val selfieFile: String, val selfieRemotePath: String, val capturedAt: String,
    val device: JsonObject, val shiftId: String? = null,
)

@Serializable
data class CheckOutPayload(
    val shiftKey: String, val lat: Double, val lng: Double, val accuracyM: Float?,
    val selfieFile: String, val selfieRemotePath: String, val capturedAt: String, val device: JsonObject,
)

@Serializable data class LocationStatePayload(val shiftKey: String, val enabled: Boolean, val at: String)
@Serializable data class PatrolStartPayload(val patrolId: String, val at: String)
@Serializable data class PatrolCompletePayload(val patrolId: String, val trail: List<List<Double>>, val photos: List<LocalPhoto>, val at: String, val notes: String? = null)
@Serializable data class TaskStartPayload(val taskId: String, val at: String)
@Serializable data class TaskCompletePayload(val taskId: String, val photo: LocalPhoto? = null, val note: String? = null, val lat: Double? = null, val lng: Double? = null, val at: String)
@Serializable data class LeaveApplyPayload(val type: String, val start: String, val end: String, val reason: String? = null)
@Serializable data class LeaveCancelPayload(val id: String)
@Serializable data class RegSelfiePayload(val file: String, val remotePath: String)

File: guard-app/src/data/payloads.ts
export const Kinds = {
  CHECK_IN: "CHECK_IN", CHECK_OUT: "CHECK_OUT", LOCATION_STATE: "LOCATION_STATE",
  PATROL_START: "PATROL_START", PATROL_COMPLETE: "PATROL_COMPLETE",
  TASK_START: "TASK_START", TASK_COMPLETE: "TASK_COMPLETE",
  LEAVE_APPLY: "LEAVE_APPLY", LEAVE_CANCEL: "LEAVE_CANCEL", REG_SELFIE: "REG_SELFIE",
} as const;

export type LocalPhoto = { file: string; remotePath: string; lat?: number | null; lng?: number | null; takenAt: string };
export type CheckInPayload = { shiftKey: string; guardId: string; siteId: string; lat: number; lng: number; accuracyM: number | null; selfieFile: string; selfieRemotePath: string; capturedAt: string; device: Record<string, unknown>; shiftId: string | null };
export type CheckOutPayload = { shiftKey: string; lat: number; lng: number; accuracyM: number | null; selfieFile: string; selfieRemotePath: string; capturedAt: string; device: Record<string, unknown> };
export type LocationStatePayload = { shiftKey: string; enabled: boolean; at: string };
export type PatrolStartPayload = { patrolId: string; at: string };
export type PatrolCompletePayload = { patrolId: string; trail: [number, number][]; photos: LocalPhoto[]; at: string; notes: string | null };
export type TaskStartPayload = { taskId: string; at: string };
export type TaskCompletePayload = { taskId: string; photo: LocalPhoto | null; note: string | null; lat: number | null; lng: number | null; at: string };
export type LeaveApplyPayload = { type: string; start: string; end: string; reason: string | null };
export type LeaveCancelPayload = { id: string };
export type RegSelfiePayload = { file: string; remotePath: string };
